File: core/src/main/java/com/ustadmobile/core/controller/RoleAssignmentDetailPresenter.java
package com.ustadmobile.core.controller;

import com.ustadmobile.core.db.UmProvider;
import com.ustadmobile.core.db.dao.ClazzDao;
import com.ustadmobile.core.db.dao.EntityRoleDao;
import com.ustadmobile.core.db.dao.LocationDao;
import com.ustadmobile.core.db.dao.PersonDao;
import com.ustadmobile.core.db.dao.PersonGroupDao;
import com.ustadmobile.core.db.dao.RoleDao;
import com.ustadmobile.core.impl.UmAccountManager;
import com.ustadmobile.core.db.UmAppDatabase;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.core.db.UmLiveData;
import com.ustadmobile.core.impl.UstadMobileSystemImpl;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Hashtable;
import java.util.List;

import com.ustadmobile.core.view.RoleAssignmentDetailView;
import com.ustadmobile.lib.db.entities.Clazz;
import com.ustadmobile.lib.db.entities.EntityRole;
import com.ustadmobile.lib.db.entities.Location;
import com.ustadmobile.lib.db.entities.Person;
import com.ustadmobile.lib.db.entities.PersonGroup;
import com.ustadmobile.lib.db.entities.Role;

import static com.ustadmobile.core.view.RoleAssignmentDetailView.ENTITYROLE_UID;


/**
 * Presenter for RoleAssignmentDetail view
 **/
public class RoleAssignmentDetailPresenter extends UstadBaseController<RoleAssignmentDetailView> {

    UmAppDatabase repository;

    private long currentEntityRoleUid = 0;
    private EntityRole originalEntityRole;
    private EntityRole updatedEntityRole;

    private UmLiveData<List<PersonGroup>> groupUmLiveData;
    private UmLiveData<List<Role>> roleUmLiveData;
    private UmLiveData<List<Location>> locationUmLiveData;
    private UmLiveData<List<Clazz>> clazzUmLiveData;
    private UmLiveData<List<Person>> personUmLiveData;

    private HashMap<Long, Integer> groupIdToPosition;
    private HashMap<Integer, Long> groupPositionToId;
    private HashMap<Long, Integer> roleIdToPosition;
    private HashMap<Integer, Long> rolePositionToId;
    private HashMap<Long, Integer> locationIdToPosition;
    private HashMap<Integer, Long> locationPositionToId;
    private HashMap<Long, Integer> clazzIdToPosition;
    private HashMap<Integer, Long> clazzPositionToId;
    private HashMap<Long, Integer> peopleIdToPosition;
    private HashMap<Integer, Long> peoplePositionToId;

    private EntityRoleDao entityRoleDao;

    private PersonGroupDao groupDao;
    private RoleDao roleDao;
    private ClazzDao clazzDao;
    private LocationDao locationDao;
    private PersonDao personDao;

    private int currentTableId;

    private String[] assigneePresets, groupPresets, rolePresets, scopePresets;

    public RoleAssignmentDetailPresenter(Object context, Hashtable arguments, RoleAssignmentDetailView view) {
        super(context, arguments, view);

        repository = UmAccountManager.getRepositoryForActiveAccount(context);
        entityRoleDao = repository.getEntityRoleDao();
        groupDao = repository.getPersonGroupDao();
        roleDao = repository.getRoleDao();
        clazzDao = repository.getClazzDao();
        locationDao = repository.getLocationDao();
        personDao = repository.getPersonDao();
        groupIdToPosition = new HashMap<>();
        groupPositionToId = new HashMap<>();
        roleIdToPosition= new HashMap<>();
        rolePositionToId = new HashMap<>();
        locationIdToPosition= new HashMap<>();
        locationPositionToId = new HashMap<>();
        clazzIdToPosition= new HashMap<>();
        clazzPositionToId = new HashMap<>();
        peopleIdToPosition= new HashMap<>();
        peoplePositionToId = new HashMap<>();

        if(arguments.containsKey(ENTITYROLE_UID)){
            currentEntityRoleUid = (long) arguments.get(ENTITYROLE_UID);
        }

    }

    @Override
    public void onCreate(Hashtable savedState) {
        super.onCreate(savedState);

        if(currentEntityRoleUid ==0){
            EntityRole newEntityRole = new EntityRole();
            newEntityRole.setErActive(false);
            entityRoleDao.insertAsync(newEntityRole, new UmCallback<Long>() {
                @Override
                public void onSuccess(Long result) {
                    initFromEntityRole(result);
                }

                @Override
                public void onFailure(Throwable exception) {
                    exception.printStackTrace();
                }
            });
        }else{
            initFromEntityRole(currentEntityRoleUid);
        }

    }

    public void updateAssigneePresets(int tableId){

        switch(tableId){
            case Clazz.TABLE_ID:
                clazzUmLiveData = clazzDao.findAllLive();
                clazzUmLiveData.observe(RoleAssignmentDetailPresenter.this,
                        RoleAssignmentDetailPresenter.this::handleAllClazzChanged);
                break;
            case Location.TABLE_ID:
                //TODO:
                break;
            case Person.TABLE_ID:
                personUmLiveData = personDao.findAllActiveLive();
                personUmLiveData.observe(RoleAssignmentDetailPresenter.this,
                        RoleAssignmentDetailPresenter.this::handleAllPersonChanged);
                break;
        }

    }

    public void initFromEntityRole(long uid){
        this.currentEntityRoleUid = uid;

        UmLiveData<EntityRole> entityRoleLiveData =
                entityRoleDao.findByUidLive(currentEntityRoleUid);
        entityRoleLiveData.observe(RoleAssignmentDetailPresenter.this,
                RoleAssignmentDetailPresenter.this::handleEntityRoleChanged);

        entityRoleDao.findByUidAsync(uid, new UmCallback<EntityRole>() {
            @Override
            public void onSuccess(EntityRole result) {
                updatedEntityRole = result;

                //Update group
                groupUmLiveData = groupDao.findAllActivePersonGroupsLive();
                groupUmLiveData.observe(RoleAssignmentDetailPresenter.this,
                        RoleAssignmentDetailPresenter.this::handleAllGroupsChanged);

                //Update roles
                roleUmLiveData = roleDao.findAllActiveRolesLive();
                roleUmLiveData.observe(RoleAssignmentDetailPresenter.this,
                        RoleAssignmentDetailPresenter.this::handleAllRolesChanged);

                long groupUid = updatedEntityRole.getErGroupUid();
                long roleUid = updatedEntityRole.getErRoleUid();
                int groupSelected = 0, roleSelected = 0;
                if(groupUid != 0 && roleUid != 0 && groupIdToPosition != null
                        && roleIdToPosition != null){
                    if(groupIdToPosition.containsKey(groupUid))
                        groupSelected = groupIdToPosition.get(groupUid);
                    if(roleIdToPosition.containsKey(roleUid))
                        roleSelected = roleIdToPosition.get(roleUid);
                }
                //Update scope and assignee
                view.updateRoleAssignmentOnView(updatedEntityRole, groupSelected,
                        roleSelected);

            }

            @Override
            public void onFailure(Throwable exception) {exception.printStackTrace();}
        });
    }


    public void handleAllGroupsChanged(List<PersonGroup> groups){
        int selectedPosition = 0;

        ArrayList<String> entityList = new ArrayList<>();
        groupIdToPosition = new HashMap<>();
        int posIter = 0;
        for(PersonGroup everyEntity: groups){
            entityList.add(everyEntity.getGroupName());
            groupIdToPosition.put(everyEntity.getGroupUid(), posIter);
            groupPositionToId.put(posIter, everyEntity.getGroupUid());
            posIter++;
        }
        groupPresets = new String[entityList.size()];
        groupPresets = entityList.toArray(groupPresets);

        if(originalEntityRole == null){
            originalEntityRole = new EntityRole();
        }
        if(originalEntityRole.getErGroupUid() != 0){
            long groupUid = originalEntityRole.getErGroupUid();
            if(groupIdToPosition.containsKey(groupUid)){
                selectedPosition = groupIdToPosition.get(groupUid);
            }
        }

        view.setGroupPresets(groupPresets, selectedPosition);
    }

    public void handleAllRolesChanged(List<Role> roles){
        int selectedPosition = 0;

        ArrayList<String> entityList = new ArrayList<>();
        roleIdToPosition = new HashMap<>();
        int posIter = 0;
        for(Role everyEntity: roles){
            entityList.add(everyEntity.getRoleName());
            roleIdToPosition.put(everyEntity.getRoleUid(), posIter);
            rolePositionToId.put(posIter, everyEntity.getRoleUid());
            posIter++;
        }
        rolePresets = new String[entityList.size()];
        rolePresets = entityList.toArray(rolePresets);

        if(originalEntityRole == null){
            originalEntityRole = new EntityRole();
        }
        if(originalEntityRole.getErRoleUid() != 0){
            long entityUid = originalEntityRole.getErRoleUid();
            if(roleIdToPosition.containsKey(entityUid)){
                selectedPosition = roleIdToPosition.get(entityUid);
            }
        }

        view.setRolePresets(rolePresets, selectedPosition);
    }

    public String[] getScopePresets() {
        return scopePresets;
    }

    public void setScopePresets(String[] scopePresets) {
        this.scopePresets = scopePresets;
    }

    public void handleAllLocationsChanged(List<Location> locations){

    }

    public void handleAllClazzChanged(List<Clazz> clazzes){
        int selectedPosition = 0;

        ArrayList<String> entityList = new ArrayList<>();
        clazzIdToPosition = new HashMap<>();
        int posIter = 0;
        for(Clazz everyEntity: clazzes){
            entityList.add(everyEntity.getClazzName());
            clazzIdToPosition.put(everyEntity.getClazzUid(), posIter);
            clazzPositionToId.put(posIter, everyEntity.getClazzUid());
            posIter++;
        }
        assigneePresets = new String[entityList.size()];
        assigneePresets = entityList.toArray(assigneePresets);

        if(originalEntityRole == null){
            originalEntityRole = new EntityRole();
        }
        if(originalEntityRole.getErEntityUid() != 0){
            long entityUid = originalEntityRole.getErEntityUid();
            if(clazzIdToPosition.containsKey(entityUid)){
                selectedPosition = clazzIdToPosition.get(entityUid);
            }
        }

        view.setAssigneePresets(assigneePresets, selectedPosition);
    }

    public void handleAllPersonChanged(List<Person> people){
        int selectedPosition = 0;

        ArrayList<String> entityList = new ArrayList<>();
        peopleIdToPosition = new HashMap<>();
        int posIter = 0;
        for(Person everyEntity: people){
            entityList.add(everyEntity.getFirstNames() + " " + everyEntity.getLastName());
            peopleIdToPosition.put(everyEntity.getPersonUid(), posIter);
            peoplePositionToId.put(posIter, everyEntity.getPersonUid());
            posIter++;
        }
        assigneePresets = new String[entityList.size()];
        assigneePresets = entityList.toArray(assigneePresets);

        if(originalEntityRole == null){
            originalEntityRole = new EntityRole();
        }
        if(originalEntityRole.getErEntityUid() != 0){
            long entityUid = originalEntityRole.getErEntityUid();
            if(peopleIdToPosition.containsKey(entityUid)){
                selectedPosition = peopleIdToPosition.get(entityUid);
            }
        }

        view.setAssigneePresets(assigneePresets, selectedPosition);
    }


    public void updateGroup(int position){

        if(groupPositionToId.containsKey(position))
            updatedEntityRole.setErGroupUid(groupPositionToId.get(position));
    }
    public void updateRole(int position){
        if(rolePositionToId.containsKey(position))
        updatedEntityRole.setErRoleUid(rolePositionToId.get(position));
    }
    public void updateScope(int position){
        currentTableId = 0;
        switch (position){
            case 0:
                currentTableId=Clazz.TABLE_ID;break;
            case 1:
                currentTableId=Person.TABLE_ID;break;
            case 2:
                currentTableId=Location.TABLE_ID;break;
        }
        updatedEntityRole.setErTableId(currentTableId);
    }

    public void updateAssignee(int position){
        long entityUid = 0;
        switch (currentTableId){
            case Clazz.TABLE_ID:
                if(clazzPositionToId.containsKey(position))
                    entityUid = clazzPositionToId.get(position);
                break;
            case Person.TABLE_ID:
                if(peoplePositionToId.containsKey(position))
                    entityUid = peoplePositionToId.get(position);
                break;
            case Location.TABLE_ID:
                if(locationPositionToId.containsKey(position))
                    entityUid = locationPositionToId.get(position);
                break;
        }
        if(entityUid != 0){
            updatedEntityRole.setErEntityUid(entityUid);
        }
    }

    public void handleEntityRoleChanged(EntityRole changedEntityRole){
        //set the og person value
        if(originalEntityRole == null)
            originalEntityRole = changedEntityRole;

        if(updatedEntityRole == null || !updatedEntityRole.equals(changedEntityRole)) {
            //update class edit views

            view.updateRoleAssignmentOnView(updatedEntityRole, 0, 0);
            //Update the currently editing class object
            updatedEntityRole = changedEntityRole;
        }
    }

    public void handleClickDone() {

        view.finish();
    }
}

File: core/src/commonMain/kotlin/com/ustadmobile/core/controller/RoleAssignmentDetailPresenter.kt
package com.ustadmobile.core.controller

import com.ustadmobile.core.db.UmAppDatabase
import com.ustadmobile.core.db.dao.*
import com.ustadmobile.core.impl.UmAccountManager
import com.ustadmobile.core.view.RoleAssignmentDetailView
import com.ustadmobile.core.view.RoleAssignmentDetailView.Companion.ENTITYROLE_UID
import com.ustadmobile.door.DoorLiveData
import com.ustadmobile.lib.db.entities.*
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Runnable
import kotlinx.coroutines.launch


/**
 * Presenter for RoleAssignmentDetail view
 */
class RoleAssignmentDetailPresenter(context: Any, arguments: Map<String, String>?, view:
RoleAssignmentDetailView) : UstadBaseController<RoleAssignmentDetailView>(context, arguments!!,
        view) {

    internal var repository: UmAppDatabase

    private var currentEntityRoleUid: Long = 0
    private var originalEntityRole: EntityRole? = null
    private var updatedEntityRole: EntityRole? = null

    private var groupUmLiveData: DoorLiveData<List<PersonGroup>>? = null
    private var roleUmLiveData: DoorLiveData<List<Role>>? = null
    private var locationUmLiveData: DoorLiveData<List<Location>>? = null
    private var clazzUmLiveData: DoorLiveData<List<Clazz>>? = null
    private var personUmLiveData: DoorLiveData<List<Person>>? = null

    private var groupIdToPosition: HashMap<Long, Int>? = null
    private val groupPositionToId: HashMap<Int, Long>
    private var roleIdToPosition: HashMap<Long, Int>? = null
    private val rolePositionToId: HashMap<Int, Long>
    private var locationIdToPosition: HashMap<Long, Int>? = null
    private val locationPositionToId: HashMap<Int, Long>
    private var clazzIdToPosition: HashMap<Long, Int>? = null
    private val clazzPositionToId: HashMap<Int, Long>
    private var peopleIdToPosition: HashMap<Long, Int>? = null
    private val peoplePositionToId: HashMap<Int, Long>

    private val entityRoleDao: EntityRoleDao

    private val groupDao: PersonGroupDao
    private val roleDao: RoleDao
    private val clazzDao: ClazzDao
    private val locationDao: LocationDao
    private val personDao: PersonDao

    private var currentTableId: Int = 0

    private var assigneePresets: Array<String>? = null
    private var groupPresets: Array<String>? = null
    private var rolePresets: Array<String>? = null
    private var scopePresets: Array<String?>? = null

    init {

        repository = UmAccountManager.getRepositoryForActiveAccount(context)
        entityRoleDao = repository.entityRoleDao
        groupDao = repository.personGroupDao
        roleDao = repository.roleDao
        clazzDao = repository.clazzDao
        locationDao = repository.locationDao
        personDao = repository.personDao
        groupIdToPosition = HashMap()
        groupPositionToId = HashMap()
        roleIdToPosition = HashMap()
        rolePositionToId = HashMap()
        locationIdToPosition = HashMap()
        locationPositionToId = HashMap()
        clazzIdToPosition = HashMap()
        clazzPositionToId = HashMap()
        peopleIdToPosition = HashMap()
        peoplePositionToId = HashMap()

        if (arguments!!.containsKey(ENTITYROLE_UID)) {
            currentEntityRoleUid = arguments!!.get(ENTITYROLE_UID)!!.toLong()
        }
    }

    fun updateGroupList(individual: Boolean) {

        if (individual) {
            //Update group
            groupUmLiveData = groupDao.findAllActivePersonPersonGroupLive()
            view.runOnUiThread(Runnable {
                groupUmLiveData!!.observe(this, this::handleAllGroupsChanged)
            })
        } else {
            //Update group
            groupUmLiveData = groupDao.findAllActiveGroupPersonGroupsLive()
            view.runOnUiThread(Runnable {
                groupUmLiveData!!.observe(this, this::handleAllGroupsChanged)
            })
        }
    }

    override fun onCreate(savedState: Map<String, String?>?) {
        super.onCreate(savedState)

        if (currentEntityRoleUid == 0L) {
            val newEntityRole = EntityRole()
            newEntityRole.erActive = false
            GlobalScope.launch {
                val result = entityRoleDao.insertAsync(newEntityRole)
                initFromEntityRole(result!!)
            }

        } else {
            initFromEntityRole(currentEntityRoleUid)
        }

    }

    fun updateAssigneePresets(tableId: Int) {

        when (tableId) {
            Clazz.TABLE_ID -> {
                clazzUmLiveData = clazzDao.findAllLive()
                clazzUmLiveData!!.observe(this, this::handleAllClazzChanged)
            }
            Location.TABLE_ID -> {
                locationUmLiveData = locationDao.findAllActiveLocationsLive()
                locationUmLiveData!!.observe(this, this::handleAllLocationsChanged)
            }
            Person.TABLE_ID -> {
                personUmLiveData = personDao.findAllActiveLive()
                personUmLiveData!!.observe(this, this::handleAllPersonChanged)
            }
        }

    }

    private fun initFromEntityRole(uid: Long) {
        this.currentEntityRoleUid = uid

        val entityRoleLiveData = entityRoleDao.findByUidLive(currentEntityRoleUid)
        var thisP = this
        view.runOnUiThread(Runnable {
            entityRoleLiveData.observe(this, this::handleEntityRoleChanged)
        })

        GlobalScope.launch {
            val result = entityRoleDao.findByUidAsync(uid)
            updatedEntityRole = result

            //Update roles
            roleUmLiveData = roleDao.findAllActiveRolesLive()

            view.runOnUiThread(Runnable {
                roleUmLiveData!!.observe(thisP, thisP::handleAllRolesChanged)
            })

            val groupUid = updatedEntityRole!!.erGroupUid
            val result2 = groupDao.findByUidAsync(groupUid)
            if (result2 != null && result2.groupPersonUid != 0L) {
                view.individualClicked()
                updateGroupList(true)
            } else {
                view.groupClicked()
                updateGroupList(false)
            }

            val roleUid = updatedEntityRole!!.erRoleUid
            var groupSelected = 0
            var roleSelected = 0
            if (groupUid != 0L && roleUid != 0L && groupIdToPosition != null
                    && roleIdToPosition != null) {
                if (groupIdToPosition!!.containsKey(groupUid))
                    groupSelected = groupIdToPosition!![groupUid]!!
                if (roleIdToPosition!!.containsKey(roleUid))
                    roleSelected = roleIdToPosition!![roleUid]!!
            }
            //Update scope and assignee
            view.updateRoleAssignmentOnView(updatedEntityRole!!, groupSelected,
                    roleSelected)
        }
    }


    private fun handleAllGroupsChanged(groups: List<PersonGroup>?) {
        var selectedPosition = 0

        val entityList = ArrayList<String>()
        groupIdToPosition = HashMap()
        var posIter = 0
        for (everyEntity in groups!!) {
            entityList.add(everyEntity.groupName!!)
            groupIdToPosition!![everyEntity.groupUid] = posIter
            groupPositionToId[posIter] = everyEntity.groupUid
            posIter++
        }
        groupPresets = entityList.toTypedArray()

        if (originalEntityRole == null) {
            originalEntityRole = EntityRole()
        }
        if (originalEntityRole!!.erGroupUid != 0L) {
            val groupUid = originalEntityRole!!.erGroupUid
            if (groupIdToPosition!!.containsKey(groupUid)) {
                selectedPosition = groupIdToPosition!![groupUid]!!
            }
        }

        view.setGroupPresets(groupPresets!!, selectedPosition)
    }

    private fun handleAllRolesChanged(roles: List<Role>?) {
        var selectedPosition = 0

        val entityList = ArrayList<String>()
        roleIdToPosition = HashMap()
        var posIter = 0
        for (everyEntity in roles!!) {
            entityList.add(everyEntity.roleName!!)
            roleIdToPosition!![everyEntity.roleUid] = posIter
            rolePositionToId[posIter] = everyEntity.roleUid
            posIter++
        }
        //rolePresets = arrayOfNulls(entityList.size)
        rolePresets = entityList.toTypedArray()

        if (originalEntityRole == null) {
            originalEntityRole = EntityRole()
        }
        if (originalEntityRole!!.erRoleUid != 0L) {
            val entityUid = originalEntityRole!!.erRoleUid
            if (roleIdToPosition!!.containsKey(entityUid)) {
                selectedPosition = roleIdToPosition!![entityUid]!!
            }
        }

        view.setRolePresets(rolePresets!!, selectedPosition)
    }

    fun setScopePresets(scopePresets: Array<String?>) {
        this.scopePresets = scopePresets
    }

    private fun handleAllLocationsChanged(locations: List<Location>?) {
        var selectedPosition = 0

        val entityList = ArrayList<String>()
        locationIdToPosition = HashMap()
        var posIter = 0
        for (everyEntity in locations!!) {
            entityList.add(everyEntity.title!!)
            locationIdToPosition!![everyEntity.locationUid] = posIter
            locationPositionToId[posIter] = everyEntity.locationUid
            posIter++
        }
        //assigneePresets = arrayOfNulls(entityList.size)
        assigneePresets = entityList.toTypedArray()
        if (originalEntityRole == null) {
            originalEntityRole = EntityRole()
        }
        if (originalEntityRole!!.erEntityUid != 0L) {
            val entityUid = originalEntityRole!!.erEntityUid
            if (locationIdToPosition!!.containsKey(entityUid)) {
                selectedPosition = locationIdToPosition!![entityUid]!!
            }
        }

        view.setAssigneePresets(assigneePresets!!, selectedPosition)
    }

    private fun handleAllClazzChanged(clazzes: List<Clazz>?) {
        var selectedPosition = 0

        val entityList = ArrayList<String>()
        clazzIdToPosition = HashMap()
        var posIter = 0
        for (everyEntity in clazzes!!) {
            entityList.add(everyEntity.clazzName!!)
            clazzIdToPosition!![everyEntity.clazzUid] = posIter
            clazzPositionToId[posIter] = everyEntity.clazzUid
            posIter++
        }
        //assigneePresets = arrayOfNulls(entityList.size)
        assigneePresets = entityList.toTypedArray()

        if (originalEntityRole == null) {
            originalEntityRole = EntityRole()
        }
        if (originalEntityRole!!.erEntityUid != 0L) {
            val entityUid = originalEntityRole!!.erEntityUid
            if (clazzIdToPosition!!.containsKey(entityUid)) {
                selectedPosition = clazzIdToPosition!![entityUid]!!
            }
        }

        view.setAssigneePresets(assigneePresets!!, selectedPosition)
    }

    private fun handleAllPersonChanged(people: List<Person>?) {
        var selectedPosition = 0

        val entityList = ArrayList<String>()
        peopleIdToPosition = HashMap()
        var posIter = 0
        for (everyEntity in people!!) {
            entityList.add(everyEntity.firstNames + " " + everyEntity.lastName)
            peopleIdToPosition!![everyEntity.personUid] = posIter
            peoplePositionToId[posIter] = everyEntity.personUid
            posIter++
        }
        //assigneePresets = arrayOfNulls(entityList.size)
        assigneePresets = entityList.toTypedArray()

        if (originalEntityRole == null) {
            originalEntityRole = EntityRole()
        }
        if (originalEntityRole!!.erEntityUid != 0L) {
            val entityUid = originalEntityRole!!.erEntityUid
            if (peopleIdToPosition!!.containsKey(entityUid)) {
                selectedPosition = peopleIdToPosition!![entityUid]!!
            }
        }

        view.setAssigneePresets(assigneePresets!!, selectedPosition)
    }


    fun updateGroup(position: Int) {

        if (groupPositionToId.containsKey(position))
            updatedEntityRole!!.erGroupUid = groupPositionToId[position]!!
    }

    fun updateRole(position: Int) {
        if (rolePositionToId.containsKey(position))
            updatedEntityRole!!.erRoleUid = rolePositionToId[position]!!
    }

    fun updateScope(position: Int) {
        currentTableId = 0
        when (position) {
            0 -> {
                currentTableId = Clazz.TABLE_ID
                view.setScopeAndAssigneeSelected(currentTableId)
            }
            1 -> {
                currentTableId = Person.TABLE_ID
                view.setScopeAndAssigneeSelected(currentTableId)
            }
            2 -> {
                currentTableId = Location.TABLE_ID
                view.setScopeAndAssigneeSelected(currentTableId)
            }
        }
        updatedEntityRole!!.erTableId = currentTableId
    }

    fun updateAssignee(position: Int) {
        var entityUid: Long = 0
        when (currentTableId) {
            Clazz.TABLE_ID -> if (clazzPositionToId.containsKey(position))
                entityUid = clazzPositionToId[position]!!
            Person.TABLE_ID -> if (peoplePositionToId.containsKey(position))
                entityUid = peoplePositionToId[position]!!
            Location.TABLE_ID -> if (locationPositionToId.containsKey(position))
                entityUid = locationPositionToId[position]!!
        }
        if (entityUid != 0L) {
            updatedEntityRole!!.erEntityUid = entityUid
        }
    }

    private fun handleEntityRoleChanged(changedEntityRole: EntityRole?) {
        //set the og person value
        if (originalEntityRole == null)
            originalEntityRole = changedEntityRole

        if (updatedEntityRole == null || updatedEntityRole != changedEntityRole) {
            //update class edit views
            //Update the currently editing class object
            updatedEntityRole = changedEntityRole

            view.updateRoleAssignmentOnView(updatedEntityRole!!, 0, 0)
        }
    }

    fun handleClickDone() {
        updatedEntityRole!!.erActive = true
        GlobalScope.launch {
            entityRoleDao.updateAsync(updatedEntityRole!!)
            view.finish()
        }
    }
}
